core/channel-upgrade.go

```diff
@@ -0,0 +1,719 @@
+			return nil, err
+		} else if complete {
+			dstAction = UPGRADE_ACTION_OPEN
+		} else if timedout, err := upgradeAlreadyTimedOut(srcCtxFinalized, src, dstChanUpg); err != nil {
+			logger.Error("failed to check if the upgrade on the src side has already timed out", err)
+			return nil, err
+		} else if timedout {
+			dstAction = UPGRADE_ACTION_TIMEOUT
+		} else {
+			dstAction = UPGRADE_ACTION_CANCEL_FLUSHCOMPLETE
+		}
+	case srcState == UPGRADE_STATE_FLUSHCOMPLETE && dstState == UPGRADE_STATE_INIT:
+		if complete, err := upgradeAlreadyComplete(dstChan, srcCtxFinalized, src, srcChanUpg); err != nil {
+			logger.Error("failed to check if the upgrade on the dst side has already completed", err)
+			return nil, err
+		} else if complete {
+			srcAction = UPGRADE_ACTION_OPEN
+		} else if timedout, err := upgradeAlreadyTimedOut(dstCtxFinalized, dst, srcChanUpg); err != nil {
+			logger.Error("failed to check if the upgrade on the dst side has already timed out", err)
+			return nil, err
+		} else if timedout {
+			srcAction = UPGRADE_ACTION_TIMEOUT
+		} else {
+			srcAction = UPGRADE_ACTION_CANCEL_FLUSHCOMPLETE
+		}
+	case srcState == UPGRADE_STATE_FLUSHING && dstState == UPGRADE_STATE_FLUSHING:
+		if timedout, err := upgradeAlreadyTimedOut(srcCtxFinalized, src, dstChanUpg); err != nil {
+			logger.Error("failed to check if the upgrade on the src side has already timed out", err)
+			return nil, err
+		} else if timedout {
+			dstAction = UPGRADE_ACTION_TIMEOUT
+		}
+		if timedout, err := upgradeAlreadyTimedOut(dstCtxFinalized, dst, srcChanUpg); err != nil {
+			logger.Error("failed to check if the upgrade on the dst side has already timed out", err)
+			return nil, err
+		} else if timedout {
+			srcAction = UPGRADE_ACTION_TIMEOUT
+		}
+
+		// if either chain has already timed out, never execute chanUpgradeConfirm
+		if srcAction == UPGRADE_ACTION_TIMEOUT || dstAction == UPGRADE_ACTION_TIMEOUT {
+			break
+		}
+
+		if completable, err := src.QueryCanTransitionToFlushComplete(srcCtxFinalized); err != nil {
+			logger.Error("failed to check if the src channel can transition to FLUSHCOMPLETE", err)
+			return nil, err
+		} else if completable {
+			srcAction = UPGRADE_ACTION_CONFIRM
+		}
+		if completable, err := dst.QueryCanTransitionToFlushComplete(dstCtxFinalized); err != nil {
+			logger.Error("failed to check if the dst channel can transition to FLUSHCOMPLETE", err)
+			return nil, err
+		} else if completable {
+			dstAction = UPGRADE_ACTION_CONFIRM
+		}
+	case srcState == UPGRADE_STATE_FLUSHING && dstState == UPGRADE_STATE_FLUSHCOMPLETE:
+		if timedout, err := upgradeAlreadyTimedOut(srcCtxFinalized, src, dstChanUpg); err != nil {
+			logger.Error("failed to check if the upgrade on the src side has already timed out", err)
+			return nil, err
+		} else if timedout {
+			dstAction = UPGRADE_ACTION_TIMEOUT
+		} else if completable, err := src.QueryCanTransitionToFlushComplete(srcCtxFinalized); err != nil {
```

Comment:
Can we use `srcCtxLatest` instead of `srcCtxFinalized`? In my understanding, the src chain does not need to wait for own state to be finalized for srcAction.

Reply:
@bluele fixed!: https://github.com/hyperledger-labs/yui-relayer/pull/143/commits/a9f0af686251836f0f4a7889b7965f840fb31ff0